### ISSUE TITLE
bug(FocusMonitor): FocusMonitor retains detached DOM-nodes and Event listeners

### ISSUE THREAD

**Issue by JonWallsten** [angular/components#23905]
### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### Description

FocusMonitor seems to retain HTML elements and Event listeners causing GC issues
![image](https://user-images.githubusercontent.com/5993909/140480115-2c030ec0-f071-42ad-82e3-6abff57f90a4.png)
![image](https://user-images.githubusercontent.com/5993909/140479989-99aa029c-8de2-4145-9b93-615482157f4c.png)

Since there's not pointers to my own internal code and the FocusMonitor is used internally it's hard for me to fix it in my code. 

### Reproduction

Steps to reproduce:
1. The project is unfortunately big and private, if this cannot be fixed any other way I can look into trying to create a repro.

### Expected Behavior

FocusMonitor should not retain detached DOM-nodes and Event listeners.
I guess it's hard to make sure that all components are removed in a safe way every single time. Maybe the FocusMonitor should remove detached nodes automatically with an interval or similar?

### Actual Behavior

When components register it's element in the FocusMonitor and it get's removed improperly the reference stay and can't be garbage collected.

### Environment

Angular CLI: 12.2.10        
Node: 14.18.1
Package Manager: npm 6.14.15
OS: win32 x64

Angular: 12.2.10
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1202.10
@angular-devkit/build-optimizer   0.1202.10
@angular-devkit/core              12.2.10
@angular-devkit/schematics        12.2.10
@angular/cdk                      12.2.9
@angular/material                 12.2.9
@ngtools/webpack                  12.2.10
@schematics/angular               12.2.10
ng-packagr                        12.2.3
rxjs                              7.4.0
typescript                        4.4.4
webpack                           5.58.2

**Comment by crisbeto** [comment]
We only retain elements that you've called `FocusMonitor.monitor` on. You should make sure that you call `stopMonitoring` when the element is destroyed. Detecting detached elements on our end can be slow, because we'd basically have to poll the DOM to check if the element is still attached.

**Comment by JonWallsten** [comment]
Unfortunately I don't have any references to FocusMonitor in my code base. I think this has something to do with material components. I saw some hints when digging around in the memory dump.

**Comment by JonWallsten** [comment]
What would generally speaking causing components to be removed without going trough the ngOnDestroy life cycle hook?

**Comment by crisbeto** [comment]
That can happen if you remove the elements from the DOM "manually" (e.g. by calling `parentNode.removeChild(element)`).

**Comment by JonWallsten** [comment]
That is not something I do. How would you investigate this? I can spend some time looking for the issue.

**Comment by crisbeto** [comment]
If you're doing everything through Angular this shouldn't really happen. It can be caused if you're integrating with something like a jQuery plugin or by a browser extension. Also have you noticed this happening after a specific action? It would also be useful to know which Material modules you're importing.

**Comment by JonWallsten** [comment]
They only tool we're using is Chrome DevTools. No extensions. It happens when I close a container with hundreds of child element in different levels. It's a complex form application with tables and trees. Lot's of material stuff, also where they are wrapped by our own components. I'll take another look and see if I find anything useful and then reply here. I'll tag you so you get a notification. Thanks!

**Comment by crisbeto** [comment]
It's worth noting that the dev tools can cause leaks as well. It won't happen if the tools are closed, but it's possible if you have a `console.log` somewhere that logs out a DOM node.

**Comment by JonWallsten** [comment]
Yeah, I've seen the cases where DevTools is responsible. You can usually see it clearly in the dump though since it has references to the DevTools in the dump. I think I even saw some in the dump above. But not for those particular detached elements.

**Comment by amysorto** [comment]
Closing this issue due to no recent activity, if this is still an issue please create a new issue.
